src/Service/MemoryInfoHandler.cpp

```diff
@@ -49,26 +54,64 @@ void MemoryInfoHandler::Stop() {
 
 void MemoryInfoHandler::OnCGroupMemoryUsage(
     orbit_grpc_protos::CGroupMemoryUsage cgroup_memory_usage) {
-  orbit_grpc_protos::ProducerCaptureEvent event;
-  *event.mutable_cgroup_memory_usage() = std::move(cgroup_memory_usage);
-  producer_event_processor_->ProcessEvent(orbit_grpc_protos::kMemoryInfoProducerId,
-                                          std::move(event));
+  uint64_t sampling_window_id = GetSamplingWindowId(cgroup_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_cgroup_memory_usage() =
+      std::move(cgroup_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
 }
 
 void MemoryInfoHandler::OnProcessMemoryUsage(
     orbit_grpc_protos::ProcessMemoryUsage process_memory_usage) {
-  orbit_grpc_protos::ProducerCaptureEvent event;
-  *event.mutable_process_memory_usage() = std::move(process_memory_usage);
-  producer_event_processor_->ProcessEvent(orbit_grpc_protos::kMemoryInfoProducerId,
-                                          std::move(event));
+  uint64_t sampling_window_id = GetSamplingWindowId(process_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_process_memory_usage() =
+      std::move(process_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
 }
 
 void MemoryInfoHandler::OnSystemMemoryUsage(
     orbit_grpc_protos::SystemMemoryUsage system_memory_usage) {
+  uint64_t sampling_window_id = GetSamplingWindowId(system_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_system_memory_usage() =
+      std::move(system_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
+}
+
+uint64_t MemoryInfoHandler::GetSamplingWindowId(uint64_t sample_timestamp_ns) const {
+  return static_cast<uint64_t>(
+      std::round(static_cast<double>(sample_timestamp_ns - sampling_start_timestamp_ns_) /
+                 sampling_period_ns_));
+}
```

Comment:
This can be a static free function - doesn't need to be a member function.

Reply:
I'm also not sure about the sampling window calculation. To me it would be more logical to put the decision boundary in the middle between to sampling periods. But here it seems to be right between the sampling periods.

This might be fine if samples can only be late, but never early. And if we cannot get out-of-sync and can have no error in the start timestamp. In this case this deserves a comment.

Reply:
`std::round` will find the nearest integer, which means we are putting *the decision boundary* in the middle between two sampling periods. For example, if the sampling period is 1 second, then:
* timestamp 0.99 ==> `sampling_window_id` is 1;
* timestamp 1.1 ==> `sampling_window_id` is 1;
* timestamp 1.6 ==> `sampling_window_id` is 2.

Reply:
Ah yes. Of course. Thanks for the explanation!

Reply:
```
return (sample_timestamp_ns - sampling_start_timestamp_ns_ + (sampling_period_ns_/2))/sampling_period_ns_;
```

Reply:
Changed to return ` (sample_timestamp_ns - sampling_start_timestamp_ns_ + (sampling_period_ns_/2))/sampling_period_ns_;
` and moved to `MemoryInfoListener.cpp` as a static non-member fucntion. Thanks for the suggestion!